---
# Evaluacion Sistema Spam



In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('datos-spam.txt', sep='|', skiprows=2, header=None, names=['A','Real','Predicho','D'], usecols=['Real','Predicho'])

In [3]:
df.head()

Real  Predicho
0     0         1
1     0         1
2     1         1
3     1         1
4     0         0

In [4]:
import numpy as np
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix


In [5]:
y_real = df['Real']
y_pred = df['Predicho']

In [6]:
# Calculamos las métricas de evaluación
accuracy = accuracy_score(y_real, y_pred)
precision = precision_score(y_real, y_pred)
recall = recall_score(y_real, y_pred)
f1 = f1_score(y_real, y_pred)
conf_matrix = confusion_matrix(y_real, y_pred)

# Calculamos las tasas de falsos positivos y negativos
FP = conf_matrix[0,1]
FN = conf_matrix[1,0]
TN = conf_matrix[0,0]
TP = conf_matrix[1,1]
FPR = FP / (FP + TN)
FNR = FN / (FN + TP)

# Mostramos los resultados
print("Métricas de evaluación:")
print("Accuracy:", accuracy)
print("Precision:", precision)
print("Recall:", recall)
print("F1-score:", f1)
print("\nMatriz de Confusión:")
print(conf_matrix)
print("\nTasas de Falsos Positivos y Negativos:")
print("Tasa de Falsos Positivos (FPR):", FPR)
print("Tasa de Falsos Negativos (FNR):", FNR)

Métricas de evaluación:
Accuracy: 0.8
Precision: 0.7
Recall: 1.0
F1-score: 0.8235294117647058

Matriz de Confusión:
[[10  6]
 [ 0 14]]

Tasas de Falsos Positivos y Negativos:
Tasa de Falsos Positivos (FPR): 0.375
Tasa de Falsos Negativos (FNR): 0.0



---

### **Análisis de las métricas**

- **Accuracy (Exactitud) = 0.8:**  
  El modelo clasifica correctamente el 80% de los casos.

- **Precision = 0.7:**  
  De todos los casos que el modelo predijo como spam, solo el 70% realmente lo eran. Hay falsos positivos.

- **Recall = 1.0:**  
  El modelo detectó todos los casos reales de spam (no dejó pasar ninguno). No hay falsos negativos.

- **F1-score = 0.82:**  
  Es una métrica balanceada entre precisión y recall. Indica buen desempeño general.

---

### **Matriz de Confusión**
```
[[10  6]
 [ 0 14]]
```
- **10:** No spam correctamente clasificados.
- **6:** No spam clasificados erróneamente como spam (falsos positivos).
- **0:** Spam clasificados erróneamente como no spam (falsos negativos).
- **14:** Spam correctamente clasificados.

---

### **Tasas de error**
- **FPR (Falsos Positivos) = 0.375:**  
  El 37.5% de los mensajes que no son spam fueron clasificados incorrectamente como spam.

- **FNR (Falsos Negativos) = 0.0:**  
  No hubo mensajes de spam que el modelo dejara pasar como no spam.

---

### **Conclusiones**

- El modelo **no deja pasar ningún spam** (recall perfecto), lo cual es bueno si queremos evitar que el spam llegue al usuario.
- Sin embargo, **clasifica erróneamente muchos mensajes legítimos como spam** (precisión y FPR bajos), lo que puede molestar a los usuarios.
- El modelo es **muy estricto**: prefiere marcar de más antes que dejar pasar spam.
- **Recomendación:**  
  Si el costo de perder mensajes legítimos es alto, se debe ajustar el modelo para reducir falsos positivos (mejorar precisión). Si el costo de dejar pasar spam es alto, este modelo es adecuado.



---

### 1. `y_proba_log = logreg.predict_proba(X_test_scaled)[:, 1]`
- **¿Qué es?**  
  Calcula la probabilidad de que cada instancia de prueba pertenezca a la clase positiva (por ejemplo, "sí compra").
- **¿Para qué sirve?**  
  Se usa para métricas como ROC AUC y para tomar decisiones basadas en probabilidades.

---

### 2. `accuracy_score(y_test, y_pred)`
- **Exactitud (Accuracy):**  
  Es el porcentaje de predicciones correctas sobre el total de casos.
- **¿Para qué sirve?**  
  Mide qué tan bien el modelo clasifica en general, pero puede ser engañosa si las clases están desbalanceadas.

---

### 3. `confusion_matrix(y_test, y_pred)`
- **Matriz de confusión:**  
  Es una tabla que muestra cuántos casos fueron clasificados correctamente o incorrectamente en cada clase.
- **¿Para qué sirve?**  
  Permite ver los errores específicos del modelo: falsos positivos, falsos negativos, verdaderos positivos y verdaderos negativos.

---

### 4. `precision_score(y_test, y_pred)`
- **Precisión (Precision):**  
  Es la proporción de verdaderos positivos sobre el total de predicciones positivas.
- **¿Para qué sirve?**  
  Indica cuántas de las predicciones positivas realmente lo son. Es útil cuando el costo de un falso positivo es alto.

---

### 5. `recall_score(y_test, y_pred)`
- **Recall (Sensibilidad):**  
  Es la proporción de verdaderos positivos sobre el total de casos positivos reales.
- **¿Para qué sirve?**  
  Mide la capacidad del modelo para encontrar todos los casos positivos. Es útil cuando el costo de un falso negativo es alto.

---

### 6. `roc_auc_score(y_test, y_proba_log)`
- **ROC AUC:**  
  Es el área bajo la curva ROC, que mide la capacidad del modelo para distinguir entre clases.
- **¿Para qué sirve?**  
  Un valor cercano a 1 indica que el modelo separa bien las clases; 0.5 indica que no es mejor que el azar.

---